### PR TITLE
Option to give leaf nodes in polytomies a default order in Tree.ladderize.

### PR DESCRIPTION
**Pull request by davipatti**
I added this option to reduce unnecessary cross overs in tangle plots of trees that have polytomies.

### CODE REVIEW

**Comment by mmore500** [comment]
@davipatti thank you for the contribution! I’m hoping to follow up and review it this upcoming week 👍

**Comment by mmore500** [comment]
@davipatti — do you think similar functionality could be achieved by sorting the tip polytomies before the ladderization process?

File: src/dendropy/datamodel/treemodel/_tree.py [hunk 1]
```diff
@@ -3096,12 +3096,26 @@ def shuffle_taxa(self, include_internal_nodes=False, rng=None):
...
         """
+
+        if default_order:
+            order = tuple(sorted(leaf.taxon.label for leaf in self.leaf_node_iter()))
+
+        def dorder(node):  # -> int
+            if default_order:
+                try:
+                    return order.index(node.taxon.label)
+                except:
```
Comment (terrycojones) [inline src/dendropy/datamodel/treemodel/_tree.py:3114]: Hello @davipatti :-)

Using a naked `except` is almost never appropriate (e.g., if the user hits control-C during the `try` the program will not be interrupted). Catch `ValueError` instead.

File: src/dendropy/datamodel/treemodel/_tree.py [hunk 2]
```diff
@@ -3096,12 +3096,26 @@ def shuffle_taxa(self, include_internal_nodes=False, rng=None):
...
-        descending (if ``ascending`` is |False|) order in terms of the number of
-        children each child node has.
+        Sorts child nodes in ascending (if ``ascending`` is |True|) or descending (if
+        ``ascending`` is |False|) order in terms of the number of children each child
+        node has. If ``default_order`` is |True| then leaf nodes get sorted by taxon
+        labels when they are in polytomies.
         """
+
+        if default_order:
+            order = tuple(sorted(leaf.taxon.label for leaf in self.leaf_node_iter()))
```
Comment (terrycojones) [inline src/dendropy/datamodel/treemodel/_tree.py:3108]: I would make this a `dict` so as not to do linear lookups later.

```python
order = dict((name, index) for (index, name) in enumerate(sorted(...))
```

File: src/dendropy/datamodel/treemodel/_tree.py (hunk 1 above)
Comment (davipatti) [reply src/dendropy/datamodel/treemodel/_tree.py:3114]: Hello @terrycojones !

No excuses for this one...

**Comment by davipatti** [comment]
> @davipatti — do you think similar functionality could be achieved by sorting the tip polytomies before the ladderization process?

@mmore500 sorry for being slow to reply. I guess that could also work. This seemed like a more explicit approach.

File: src/dendropy/datamodel/treemodel/_tree.py [hunk 3]
```diff
@@ -3096,12 +3096,31 @@ def shuffle_taxa(self, include_internal_nodes=False, rng=None):
...
+            }
+
+        def dorder(node):  # -> int
+            if default_order:
+                try:
+                    return order[node.taxon.label]
+                except ValueError:
+                    return len(order)
+            else:
+                return 0
```
Comment (terrycojones) [inline src/dendropy/datamodel/treemodel/_tree.py:3122]: This should be `return order.get(node.taxon.label, len(order)) if default_order else 0`.

Comment (terrycojones) [reply src/dendropy/datamodel/treemodel/_tree.py:3122]: Actually, you could get rid of the `dorder` function, and just use

```python
key=lambda n: (node_desc_counts[n], order.get(node.taxon.label, len(order)) if default_order else 0)
```

in the `sort`.

Comment (davipatti) [reply src/dendropy/datamodel/treemodel/_tree.py:3122]: We can also just get rid of constructing the `order` dict and just pass the taxon label as the second item to sort on:

```python
key=lambda n: (
    node_desc_counts[n],
    n.taxon.label if default_order and hasattr(n.taxon, "label") else 0
),
```

(Also, the above naked except was catching cases where the taxon doesn't have a label, so added that here.

I guess that not all nodes must have a taxon, so there should also be `and hasattr(n, "taxon")` before checking the taxon has a label.

**Comment by mmore500** [comment]
@davipatti I ended up pushing to the wrong branch by accident, so this pull request ended up closing sooner than I intended without being able to run suggestions by.

I implemented a "reorder" function that, by default, sorts by label. I added a note to the documentation suggesting that ladderization sort is stable, so to control order within ladderization, call reorder before ladderizing. Let me know if this covers your use case.

**Pull request merged**